### PR TITLE
[AMDGPU][CodeGen] Do not rematerialize divergent registers with convergent users

### PR DESCRIPTION
**Pull request by lucas-rami**
A convergent user (e.g., `V_READLANE`) of a divergent register may read lanes of its operand that are inactive under the current EXEC mask; the value in those lanes depends on the EXEC mask in effect at the definition. Moving the definition into the user's block can place it under a different (narrower) EXEC mask, changing what the user observes in inactive lanes and producing incorrect results. At this pipeline stage EXEC is constant inside a block so a rematerialization within the same block is always safe.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir [hunk 1]
```diff
@@ -13563,3 +13563,55 @@ body:             |
...
+    ; GFX908-GCNTRACKERS-NEXT: S_ENDPGM 0
+    %saved_exec:sreg_64 = S_MOV_B64 $exec
+    %full_exec:sreg_64 = S_MOV_B64 -1
+    $exec = S_MOV_B64 %full_exec
+
+    %pressure:vreg_1024 = IMPLICIT_DEF
+    %candidate:vgpr_32 = V_MOV_B32_e32 42, implicit $exec
+
+    %narrow_exec:sreg_64 = S_MOV_B64 1
+    $exec = S_MOV_B64 %narrow_exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir]: This is supposed to be illegal in the middle of the block, this should be a realistic example

File: llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp [hunk 2]
```diff
@@ -1593,6 +1593,15 @@ bool PreRARematStage::initGCNSchedStage() {
                [](const MachineInstr *DefMI) { return DefMI->isConvergent(); }))
       continue;
 
+    // A convergent user (e.g., V_READLANE*) may observe the definition's lanes
+    // whose contents depend on the EXEC mask in effect at the def. Moving the
+    // def into the use's region can change EXEC across the def and thus alter
+    // those lanes, so prevent rematerialization in that case.
+    if (any_of(Users, [](const MachineInstr *UserMI) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp]: This would only apply to VALU defs, but something also feels off about checking the users

**Comment by gbaraldi** [comment]
Is this related to https://github.com/llvm/llvm-project/issues/222368?

File: llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir (hunk 1 above)
Comment (lucas-rami) [reply llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir]: I made a more realistic example without an EXEC change. We can't reason about EXEC across basic blocks AFAIU so an explicit EXEC change is even irrelevant in the test; the logic should reject the remat anyway.

File: llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp (hunk 2 above)
Comment (lucas-rami) [reply llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp]: I added a check so that we check user convergence for divergent registers only, and allowed convergent users if they are in the same block as the original definition.

> but something also feels off about checking the users

Do you think it's a restriction that is too conservative?

**Comment by lucas-rami** [comment]
> Is this related to https://github.com/llvm/llvm-project/issues/222368?

This looks like the same sort of issue conceptually (rematerialization to an invalid position) but this patch won't resolve it unfortunately; this deals with rematerialization in the scheduler not the register allocator.

**Comment by qcolombet** [comment]
> Convergent users of rematerializable registers may observe register lanes that are active at the original definition point but inactive at the rematerialization point, and will thus read incorrect values if we allow the rematerialization to occur

I understand what you're describing but at the same time I feel that every optimization in LLVM will happily move definitions like this across the control flow. Are we dropping the convergent flag on the definition?

It doesn't feel right that we need to check the users but I may be missing something (I think @arsenm as a similar concern if I understand https://github.com/llvm/llvm-project/pull/222322#discussion_r3968640488)

**Comment by lucas-rami** [comment]
> I understand what you're describing but at the same time I feel that every optimization in LLVM will happily move definitions like this across the control flow. Are we dropping the convergent flag on the definition?

I am not super familiar with all the convergence/divergence handling in LLVM so I may be missing part of the picture here. I don't think we are dropping any flag on the definition by rematerializing it.

> It doesn't feel right that we need to check the users but I may be missing something

Here is a minimal example to discuss the issue on something concrete.

```mir
bb.0:
  successors: bb.1:
  
  $exec = S_MOV_B64 -1 # All lanes active

bb.1:
  successors: bb.2:

  %candidate:vgpr_32 = V_MOV_B32_e32 0, implicit $exec # Applies on all lanes
  $exec = S_MOV_B64 1 # Only lane 0 is active from now on

bb.2:
  # <-- Without this PR, %candidate can be rematerialized here
  %lane_1_of_candidate:sreg_32 = V_READLANE_B32 %candidate, 1 # Reads 0
```

Assuming pressure conditions make it profitable, `%candidate` can currently be rematerialized at the beginning of bb.2, which would make `%lane_1_of_candidate` read an unknown value. I don't see a way to detect that situation and forbid the rematerialization without looking at the user. Or is this piece of MIR invalid in the first place?

**Comment by qcolombet** [comment]
> > I understand what you're describing but at the same time I feel that every optimization in LLVM will happily move definitions like this across the control flow. Are we dropping the convergent flag on the definition?
> 
> I am not super familiar with all the convergence/divergence handling in LLVM so I may be missing part of the picture here. I don't think we are dropping any flag on the definition by rematerializing it.

Sorry, let me be more specific, by dropping the flag I was wondering if at the LLVM IR the definition has the convergent flag but we somehow lost it when we lower to Machine IR and that's the reason of the bug.

**Comment by qcolombet** [comment]
```
  %candidate:vgpr_32 = V_MOV_B32_e32 0, implicit $exec # Applies on all lanes
  $exec = S_MOV_B64 1 # Only lane 0 is active from now on
```
It shouldn't be legal to move `%candidate` past the redef of `$exec`. I don't think it's technically a `convergent` flag issue, although we could argue that `%candidate` should be tagged `convergent`.

Why do we ignore the redef of `$exec`?

**Comment by lucas-rami** [comment]
> Sorry, let me be more specific, by dropping the flag I was wondering if at the LLVM IR the definition has the convergent flag but we somehow lost it when we lower to Machine IR and that's the reason of the bug.

Ah, understood. I will look at the LLIR for the miscompile motivating this fix and verify that no convergent flag was spuriously dropped.

> It shouldn't be legal to move %candidate past the redef of $exec. I don't think it's technically a convergent flag issue, although we could argue that %candidate should be tagged convergent.
Why do we ignore the redef of $exec?

In the "common case" where all uses of the register we wish to rematerialize are divergent it should always be ok to rematerialize inside the scheduling region / block in which all the uses are. In this case the rematerialized def and uses would fall under the same EXEC mask. So I think we can safely ignore any change to `$exec` in such cases, which is what I have been doing in all cases so far.

When at least one user is convergent, the "different basic block check" [here](https://github.com/llvm/llvm-project/pull/222322/changes#diff-0cc602cd384aebd0894c4e02f293602abf32a93d143d90269b4eb4b20cfdc8bfR1610-R1615) is a conservative way of checking whether the EXEC mask could be different between the original and rematerialized def positions.

**Comment by jayfoad** [comment]
> A convergent user (e.g., `V_READLANE`) of a divergent register may read lanes of its operand that are inactive under the current EXEC mask; the value in those lanes depends on the EXEC mask in effect at the definition.

That's not really what "convergent" means. In the IR programming model trying to read from an inactive lane should always give undefined results. Or _almost_ always... there are corner cases, e.g. if you are using the WWM intrinsics.

All the test cases in this PR are MIR. Do you have an end to end test case starting from LLVM IR?

To put it another way: where does the `V_READLANE` come from in the first place?

**Comment by ruiling** [comment]
>In the IR programming model trying to read from an inactive lane should always give undefined results.

+1 on this.

**Comment by lucas-rami** [comment]
Thanks all for reviewing this and helping me understand convergence better. I now think the patch is most likely undesirable. It also looks like the issue that motivated this fix is no longer happening / was caused by something else, so closing this.
